### PR TITLE
chore: Upgrade to DataFusion 44.0.0-rc2

### PR DESCRIPTION
**Pull request by andygrove**
## Which issue does this PR close?

N/A

## Rationale for this change
- Part of https://github.com/apache/datafusion/issues/13835

Start updating in preparation for DF 44.0.0

## What changes are included in this PR?

- Manually implement `Hash` & `PartialEq` for many expressions (due to changes in https://github.com/apache/datafusion/pull/13005)
- Other minor changes

## How are these changes tested?

Existing tests

### CODE REVIEW

**Comment by kazuyukitanimura** [comment]
> Stop running the test suite with miri 

Hmmm
Since this is not an official release yet, should we wait for the miri fix?

**Comment by andygrove** [comment]
> > Stop running the test suite with miri
> 
> Hmmm Since this is not an official release yet, should we wait for the miri fix?

Yes, I think we should. There is a PR to fix in DF: https://github.com/apache/datafusion/pull/13778

**Comment by alamb** [comment]
> I have upgraded to latest DF and reinstated the miri checks. I'll keep this in draft until the RC is available later this week.

I see the MIRI CI check fails:

https://github.com/apache/datafusion-comet/actions/runs/12455101824/job/34767108969?pr=1154

I can reproduce the failure locally on this branch with this command (needs rust nightly)
```shell
MIRIFLAGS="-Zmiri-disable-isolation" cargo +nightly miri test -p datafusion-comet -- test_unpack_dictionary_primitive
```

I have an idea of what is wrong -- I am looking into this

**Comment by alamb** [comment]
I found the issue: 
- https://github.com/apache/datafusion/pull/13778#issuecomment-2559808094

Working on a fix

**Comment by andygrove** [comment]
@viirya @kazuyukitanimura  This is ready for review now.

File: native/spark-expr/src/strings.rs [hunk 1]
```diff
@@ -141,18 +137,43 @@ make_predicate_function!(EndsWith, ends_with_dyn, ends_with_utf8_scalar_dyn);
 
 make_predicate_function!(Contains, contains_dyn, contains_utf8_scalar_dyn);
 
-#[derive(Debug, Hash)]
```
Comment (viirya) [inline native/spark-expr/src/strings.rs:144]: Why `Hash` cannot be derived now?

File: native/spark-expr/src/strings.rs [hunk 2]
```diff
@@ -141,18 +137,43 @@ make_predicate_function!(EndsWith, ends_with_dyn, ends_with_utf8_scalar_dyn);
...
+        self.start.hash(state);
+        self.len.hash(state);
+    }
+}
+
+impl PartialEq for SubstringExpr {
+    fn eq(&self, other: &Self) -> bool {
+        self.child.eq(&other.child) && self.start.eq(&other.start) && self.len.eq(&other.len)
+    }
+}
```
Comment (viirya) [inline native/spark-expr/src/strings.rs:159]: Why we need to provide `PartialEq` implementation now?

**Review by viirya — COMMENTED** [review]
Looks like most changes are around `PartialEq`, `Eq` and `Hash`. Is any particular change in DataFusion for this?

**Comment by andygrove** [comment]
> Looks like most changes are around `PartialEq`, `Eq` and `Hash`. Is any particular change in DataFusion for this?

Yes, this is due to https://github.com/apache/datafusion/pull/13005

**Comment by viirya** [comment]
Looks good. I re-triggered a failed Spark Hive test. It should be just flaky one.

**Review by viirya — APPROVED** [review]

**Comment by andygrove** [comment]
Thanks for the review @viirya

**Pull request merged**